### PR TITLE
ccip - graceful shutdown of chain readers and contract writers

### PR DESCRIPTION
**Pull request by dimkouv**
Graceful shutdown of chain readers and contract writers.

### CODE REVIEW

**Review by makramkd — DISMISSED** [review]

File: core/capabilities/ccip/oraclecreator/plugin.go [hunk 1]
```diff
@@ -203,7 +204,22 @@ func (i *pluginOracleCreator) Create(donID uint32, config cctypes.OCR3ConfigWith
 	if err != nil {
 		return nil, err
 	}
-	return oracle, nil
+
+	closeFunc := func() error {
```
Comment (mateusz-sekara) [inline core/capabilities/ccip/oraclecreator/plugin.go]: Wouldn't it be simpler to just pass all the closeable objects as slice and let the `wrappedOracle` do its job? This lambda function is a bit hacky to me. I mean something like this.

```go
closeableResources := make([]io.Closer, 0)
closeableResources = append(closeableResources, maps.Values(chainReaders))
closeableResources = append(closeableResources, maps.Values(chainWriters))

return newWrappedOracle(oracle, closeableResource), nil
```

And then iterate through that slice in the wrapped oracle? This way you can pass there anything that implements `Close()`

File: core/capabilities/ccip/oraclecreator/plugin.go [hunk 2]
```diff
@@ -203,7 +204,22 @@ func (i *pluginOracleCreator) Create(donID uint32, config cctypes.OCR3ConfigWith
...
+
+		for _, cr := range contractReaders {
+			errs = append(errs, cr.Close())
+		}
+
+		for _, cw := range chainWriters {
+			errs = append(errs, cw.Close())
+		}
+
+		return multierr.Combine(errs...)
```
Comment (jmank88) [inline core/capabilities/ccip/oraclecreator/plugin.go]: We're trying to move away from `multierr`. How about using the std lib?
```suggestion
	closeFunc := func() (err error) {
		for _, cr := range contractReaders {
			err = errors.Join(err, cr.Close())
		}

		for _, cw := range chainWriters {
			err = errors.Join(err, cw.Close())
		}

		return
```
Or, you can close these in parallel with via helpers:
```suggestion
	closeFunc := func() error {
		return services.CloseAll(slices.Concat[io.Closer](contractReaders, chainWriters)...)
```

File: core/capabilities/ccip/oraclecreator/wrapped_oracle.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		errs = append(errs, fmt.Errorf("close base oracle: %w", err))
+	}
+
+	if o.afterCloseFunc != nil {
+		if err := o.afterCloseFunc(); err != nil {
+			errs = append(errs, fmt.Errorf("after close func: %w", err))
+		}
+	}
+
+	return multierr.Combine(errs...)
```
Comment (jmank88) [inline core/capabilities/ccip/oraclecreator/wrapped_oracle.go]:
```suggestion
	return services.CloseAll(o.baseOracle, o.afterCloseFunc)
```

File: core/capabilities/ccip/oraclecreator/plugin.go (hunk 1 above)
Comment (jmank88) [reply core/capabilities/ccip/oraclecreator/plugin.go]: Yeah this better than my suggestion, but you should still consider using `services.CloseAll` internally.

Comment (dimkouv) [reply core/capabilities/ccip/oraclecreator/plugin.go]: Applied the pattern that mateusz suggested.
And used `services.CloseAll` for services that can be closed concurrently.

**Comment by dimkouv** [comment]
Thank you for your reviews @makramkd @mateusz-sekara and @jmank88 .

Resolved all your comments.

**Review by jmank88 — APPROVED** [review]

**Review by mateusz-sekara — APPROVED** [review]

**Pull request merged**